### PR TITLE
Adds fallback layer for multi cluster code in External mode

### PR DESCRIPTION
**Pull request by bipuladh**

### CODE REVIEW

File: packages/ocs/dashboards/persistent-external/utilization-card.tsx [hunk 1]
```diff
@@ -29,7 +30,9 @@ export const UtilizationContent: React.FC = () => {
   const { systemFlags } = useODFSystemFlagsSelector();
 
   // name of created StorageClasses are prefix by StorageCluster name
-  const storageClassNamePrefix = systemFlags[clusterNs]?.ocsClusterName;
+  // name of created StorageClasses are prefix by StorageCluster name
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/utilization-card.tsx:33]: nits
```suggestion
```

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export const storageClusterResource: WatchK8sResource = {
+  isList: true,
+  kind: referenceForModel(OCSStorageClusterModel),
+};
+
+const fsProvisionerName = '.cephfs.csi.ceph.com';
+
+const useFallbackHook = (storageClusterName) => {
+  const [storageClasses, storageClassesLoaded, storageClassLoadError] =
+    useK8sWatchResource<StorageClassResourceKind[]>(storageClusterResource);
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:18]: don't u wanna pass StorageClass resource instead...
```suggestion
    useK8sWatchResource<StorageClassResourceKind[]>(storageClassResource);
```

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  const provisioners = storageClasses.map((sc) => sc.provisioner);
+  const fileSystemProvisioner = provisioners.find((item) =>
+    item.includes(fsProvisionerName)
+  );
+  const clusterName = fileSystemProvisioner?.split(fsProvisionerName)[0];
+  return storageClassesLoaded && !storageClassLoadError
+    ? clusterName
+    : 'ocs-storagecluster';
```
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:30]: also, during multiple StorageClusyter epic we removed all explicit reference to term `ocs-external-storagecluster` & `ocs-storagecluster` (as name can be configurable in future)... can we read this value from some CR instead ??

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  WatchK8sResource,
+  useK8sWatchResource,
+} from '@openshift-console/dynamic-plugin-sdk';
+
+export const storageClusterResource: WatchK8sResource = {
+  isList: true,
+  kind: referenceForModel(OCSStorageClusterModel),
+};
+
+const fsProvisionerName = '.cephfs.csi.ceph.com';
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:14]: nits
```suggestion
const fsProvisionerPostFix = '.cephfs.csi.ceph.com';
```

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 2 above)
Comment (GowthamShanmugam) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts]: My confusion is for a storageClass watcher why storageClusterModel is used? correct me if I am missing anything about multi-cluster.

File: packages/ocs/dashboards/persistent-external/utilization-card.tsx [hunk 5]
```diff
@@ -29,7 +30,9 @@ export const UtilizationContent: React.FC = () => {
...
+  const storageClassName = systemFlags?.[clusterNs]?.ocsClusterName;
+  const storageClassNamePrefix = useFallbackHook(storageClassName);
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/utilization-card.tsx:35]: nits
```suggestion
  const ocsClusterName = systemFlags?.[clusterNs]?.ocsClusterName;
  const storageClassNamePrefix = useFallbackHook(ocsClusterName);
```

File: packages/ocs/dashboards/persistent-external/breakdown-card.tsx [hunk 6]
```diff
@@ -51,7 +52,8 @@ export const BreakdownCard: React.FC = () => {
   const { systemFlags } = useODFSystemFlagsSelector();
 
   // name of created StorageClasses are prefix by StorageCluster name
-  const storageClassNamePrefix = systemFlags[clusterNs]?.ocsClusterName;
+  const storageClassName = systemFlags?.[clusterNs]?.ocsClusterName;
+  const storageClassNamePrefix = useFallbackHook(storageClassName);
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/breakdown-card.tsx:56]: nits
```suggestion
  const ocsClusterName = systemFlags?.[clusterNs]?.ocsClusterName;
  const storageClassNamePrefix = useFallbackHook(ocsClusterName);
```

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:30]: no they use `ocs-storagecluster` as default for it client mode

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import { referenceForModel } from '@odf/shared/utils';
+import {
+  WatchK8sResource,
+  useK8sWatchResource,
+} from '@openshift-console/dynamic-plugin-sdk';
+
+export const storageClusterResource: WatchK8sResource = {
+  isList: true,
+  kind: referenceForModel(OCSStorageClusterModel),
+};
```
Comment (GowthamShanmugam) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts]: this watch resource variable is not used in more than one place, i would prefer to use the parameters directly on the watch

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:30]: is client/consumer a external mode cluster, they are still using `ocs-storagecluster` as a prefix ??

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+const fsProvisionerName = '.cephfs.csi.ceph.com';
+
+const useFallbackHook = (storageClusterName) => {
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:16]: can we rename the hook to maybe `useClientFallback` ?? to specify that it is only used for Client use case and not needed by ODF...

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 2 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:18]: yes. mistake

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 8 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:16]: I will add a comment explaining that instead.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 7 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts]: Keeps the hook code shorter.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 2 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts]: it's a mistake.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+const fsProvisionerPostFix = '.cephfs.csi.ceph.com';
+
+/**
+ * This hooks is for ODF Client Mode use case only.
+ * ODF client mode doesn't deploy ODF StorageCluster hence falling back to StorageClases.
+ */
+const useFallbackHook = (storageClusterName) => {
+  const [storageClasses, storageClassesLoaded, storageClassLoadError] =
+    useK8sWatchResource<StorageClassResourceKind[]>(scResource);
```
Comment (SanjalKatiyar) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:22]: this way it won't even poll when we already have `storageClusterName`...
```suggestion
   useK8sWatchResource<StorageClassResourceKind[]>(getValidWatchK8sResourceObj(scResource, !storageClusterName));
```

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts]: We already poll for storageClass in the same page. so we can live with this poll. Essentially the connection is going to be shared.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 8 above)
Comment (SanjalKatiyar) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts]: okay, thing is we are returning `ocs-storagecluster` (as edge case)... but in ODF the name will be `ocs-external-storagecluster`, so this is a bit client specific hook...

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:34]: client isn't external mode cluster.

[the remark this answers is not in the document]
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:34]: Screenshot from 2024-02-07 17-17-51

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 8 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts]: changed the name

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:34]: ack !

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 9 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:22]: this resource is already called in the same page for inventory card. so we can live with this.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  WatchK8sResource,
+  useK8sWatchResource,
+} from '@openshift-console/dynamic-plugin-sdk';
+
+const scResource: WatchK8sResource = {
+  isList: true,
+  kind: referenceForModel(StorageClassModel),
+};
+
+const fsProvisionerPostFix = '.cephfs.csi.ceph.com';
```
Comment (GowthamShanmugam) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:14]: under common.ts let's create an enum for each ceph provisioner type and reuse the same here, in this way we can replace the cephStorageProvisioners list under common also can be replaced with enums.

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 9 above)
Comment (SanjalKatiyar) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:22]: make sense then !! (I did not check the parent component)

**Comment by SanjalKatiyar** [comment]
LGTM

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 10 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:14]: Tracking at https://github.com/red-hat-storage/odf-console/issues/1215

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+const useClientFallback = (storageClusterName) => {
+  const [storageClasses, storageClassesLoaded, storageClassLoadError] =
+    useK8sWatchResource<StorageClassResourceKind[]>(scResource);
+  if (storageClusterName) {
+    return storageClusterName;
+  }
+
+  const provisioners = storageClasses.map((sc) => sc.provisioner);
+  const fileSystemProvisioner = provisioners.find((item) =>
+    item.includes(fsProvisionerPostFix)
```
Comment (GowthamShanmugam) [inline packages/ocs/dashboards/persistent-external/fallback-hook.ts:29]: shall we replace this with a util function like isCephFSProvisioner?

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 10 above)
Comment (GowthamShanmugam) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:14]: ack

**Review by GowthamShanmugam — APPROVED** [review]

File: packages/ocs/dashboards/persistent-external/fallback-hook.ts (hunk 11 above)
Comment (bipuladh) [reply packages/ocs/dashboards/persistent-external/fallback-hook.ts:29]: we can do that but I don't think we would have a very specific use case. We could probably go with isOCSProvisioner with more args. Adding it to #1215

**Comment by SanjalKatiyar** [comment]
/retitle Adds fallback layer for multi cluster code in External mode

**Pull request merged**